stable/database/files/nuosm

```diff
@@ -33,312 +39,499 @@ if [ $crashcount -ge 1 ]; then
   fi
 fi
 
-export NUOCMD DB_NAME NUODB_STORAGE_PASSWORDS_DIR
+export DB_NAME NUODB_STORAGE_PASSWORDS_DIR
 
 #=======================================
 # function - report an error and exit
 #
 function die() {
-  # cleanup
-  $NUOCMD set value --key $first_req --value '' --expected-value $HOSTNAME
-
   retval=$1
   shift
-  echo "$@"
+  error="$@"
+  [ -n "$wotIdid" ] && error="Error while ${wotIdid}: ${error}"
+  log "$error"
+
+  # cleanup
+  nuocmd set value --key $first_req --value '' --expected-value $HOSTNAME
+  nuocmd set value --key $startup_key/$DB_NAME --value '' --expected-value $HOSTNAME
+
+  # release my archive if it is locked
+  releaseLockedArchive
+
   exit $retval
 }
 
 #=======================================
-# function - perform archive restore
+# function - log messages
+#
+function log() {
+  local err="$@"
+  echo "$err" | tee -a $LOGFILE
+}
+
+#=======================================
+# function - trace workflow
+#
+function trace() {
+  wotIdid="$@"
+  [ -n "$1" -a -n "$NUODB_DEBUG" ] && log "trace: $wotIdid"
+}
+
+#=======================================
+# function - wrap a log file around so it doesn't grow infinitely
+#
+function wrapLogfile() {
+  logsize=$( du -sb $LOGFILE | grep -o '^ *[0-9]\+' )
+  maxlog=5000000
+  log "logsize=$logsize; maxlog=$maxlog"
+  if [ ${logsize:=0} -gt $maxlog ]; then
+    lines=$(wc -l $LOGFILE)
+    tail -n $(( lines / 2 )) $LOGFILE > ${LOGFILE}-new
+    rm -rf $LOGFILE
+    mv $LOGFILE-new $LOGFILE
+    log "(nuosm) log file wrapped around"
+  fi
+}
```

Comment:
How much logging do we expect this to generate that we are not just emitting to standard output?

Reply:
I see the value of having a `nuosm` persistent log which is available between container restarts. All restore/import logic will be executed only once and it's easy to lose the logging from it especially if the user tried to resolve a restore problem on their own. Of course, all this will work if `log.pesistance` is enabled for SMs.

Example restore log:
```
===========================================
logsize=1487; maxlog=5000000
Directory /var/opt/nuodb/archive/nuodb/demo exists
myArchive=0; DB=demo; hostname=sm-database-aqochm-nuodb-cluster0-demo-hotcopy-0
restore_source=20201229T225653; restore_requested=20201229T225653; path=/var/opt/nuodb/archive/nuodb/demo; atoms=78; catalogs=213
First-in = sm-database-aqochm-nuodb-cluster0-demo-hotcopy-0
I am first-in: sm-database-aqochm-nuodb-cluster0-demo-hotcopy-0 == sm-database-aqochm-nuodb-cluster0-demo-hotcopy-0
Deleting archiveId=1
Restoring 20201229T225653; existing archive directores: total 4
drwxr-xr-x 37 nuodb root 4096 Dec 29 22:57 demo
(restore) recreated /var/opt/nuodb/archive/nuodb/demo; atoms=0
Calling nuodocker to restore 20201229T225653 into /var/opt/nuodb/archive/nuodb/demo
Finished restoring /var/opt/nuodb/backup/20201229T225653 to /var/opt/nuodb/archive/nuodb/demo. Created archive with archive ID 2
Deleting my archive metadata: 0
Restored from 20201229T225653
Clearing restore credential request from raft
[2] <NO VALUE> : /var/opt/nuodb/archive/nuodb/demo @ demo [journal_path = ] [snapshot_archive_path = ] NOT_RUNNING
```

Example normal startup:
```
===========================================
logsize=44; maxlog=5000000
Created new dir /var/opt/nuodb/archive/nuodb/demo
myArchive=-1; DB=demo; hostname=sm-database-aqochm-nuodb-cluster0-demo-hotcopy-0
restore_source=; restore_requested=; path=/var/opt/nuodb/archive/nuodb/demo; atoms=0; catalogs=0
```